staging/src/k8s.io/cloud-provider/controllers/service/controller.go

```diff
@@ -659,75 +724,86 @@ func nodeReadyConditionStatus(node *v1.Node) v1.ConditionStatus {
 	return ""
 }
 
-// nodeSyncLoop handles updating the hosts pointed to by all load
+// nodeSyncInternal handles updating the hosts pointed to by all load
 // balancers whenever the set of nodes in the cluster changes.
-func (s *Controller) nodeSyncLoop() {
-	s.knownHostsLock.Lock()
-	defer s.knownHostsLock.Unlock()
+func (s *Controller) nodeSyncInternal(workers int) {
 	startTime := time.Now()
 	defer func() {
 		latency := time.Now().Sub(startTime).Seconds()
-		klog.V(4).Infof("It took %v seconds to finish nodeSyncLoop", latency)
+		klog.V(4).Infof("It took %v seconds to finish nodeSyncInternal", latency)
 		nodeSyncLatency.Observe(latency)
 	}()
 
-	newHosts, err := listWithPredicate(s.nodeLister, s.getNodeConditionPredicate())
-	if err != nil {
-		runtime.HandleError(fmt.Errorf("Failed to retrieve current set of nodes from node lister: %v", err))
-		return
-	}
-	if nodeSlicesEqualForLB(newHosts, s.knownHosts) {
+	if !s.needFullSyncAndUnmark() {
 		// The set of nodes in the cluster hasn't changed, but we can retry
 		// updating any services that we failed to update last time around.
-		s.servicesToUpdate = s.updateLoadBalancerHosts(s.servicesToUpdate, newHosts)
+		s.servicesToUpdate = s.updateLoadBalancerHosts(s.servicesToUpdate, workers)
 		return
 	}
-
-	klog.V(2).Infof("Detected change in list of current cluster nodes. New node set: %v",
-		nodeNames(newHosts))
+	klog.V(2).Infof("Syncing backends for all LB services.")
 
 	// Try updating all services, and save the ones that fail to try again next
 	// round.
 	s.servicesToUpdate = s.cache.allServices()
 	numServices := len(s.servicesToUpdate)
-	s.servicesToUpdate = s.updateLoadBalancerHosts(s.servicesToUpdate, newHosts)
+	s.servicesToUpdate = s.updateLoadBalancerHosts(s.servicesToUpdate, workers)
 	klog.V(2).Infof("Successfully updated %d out of %d load balancers to direct traffic to the updated set of nodes",
 		numServices-len(s.servicesToUpdate), numServices)
+}
+
+// nodeSyncService syncs the nodes for one load balancer type service
+func (s *Controller) nodeSyncService(svc *v1.Service) bool {
+	if svc == nil || !wantsLoadBalancer(svc) {
```

Comment:
The logic of removing LBs etc. is triggered by the changes of service objects, right?

Reply:
Yes. NodeSync is only updating LB hosts.